File: src/com/timepath/hl2/io/demo/Message.java
package com.timepath.hl2.io.demo;

import com.timepath.DataUtils;
import com.timepath.Pair;
import com.timepath.io.BitBuffer;
import java.nio.ByteBuffer;
import java.text.MessageFormat;
import java.util.LinkedList;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 *
 * @author TimePath
 */
public class Message {

    private static final Logger LOG = Logger.getLogger(Message.class.getName());

    public ByteBuffer data;
    
    public List<Pair<Object, Object>> meta = new LinkedList<Pair<Object, Object>>();

    /**
     * Actually 3 bytes
     */
    public final int tick;

    public final MessageType type;

    private final HL2DEM outer;

    Message(ByteBuffer buffer, final HL2DEM outer) {
        this.outer = outer;
        int b = buffer.get();
        type = MessageType.get(b);
        if(type == null) {
            LOG.log(Level.SEVERE, "Unknown demo message type encountered: {0}", b);
        }
        tick = buffer.getShort() + (buffer.get() << 16);
        if(type != MessageType.Stop) {
            buffer.get();
        }
        LOG.log(Level.FINE, "{0} at tick {1} ({2}), {3} remaining bytes",
                new Object[] {type, tick, buffer.position(), buffer.remaining()});
    }

    @Override
    public String toString() {
        return MessageFormat.format("{0}, tick {1}, {2} bytes", type, tick, data != null ? data.limit() : 0);
    }

    void parse() {
        if(data == null) {
            return;
        }
        switch(type) {
            case Signon:
            case Packet: {
                BitBuffer bb = new BitBuffer(data);
                while(true) {
                    if(bb.remaining() < 1) {
                        break;
                    }
                    int mid = (int) bb.getBits(outer.header.networkProtocol >= 16 ? 6 : 5);
                    Packet p = Packet.get(mid);
                    if(p == null) {
                        String str = MessageFormat.format("Unknown message type {0} at {1}", mid, tick);
                        LOG.log(Level.WARNING, str);
                        meta.add(new Pair<Object, Object>("error", str));
                        break;
                    }
                    
                    List<Pair<Object, Object>> list = new LinkedList<Pair<Object, Object>>();
                    boolean complete = false;
                    try {
                        complete = p.handler.read(bb, list, outer);
                        if(!complete) {
                            String str = MessageFormat.format("Incomplete read of {0} at {1}", p, tick);
                            LOG.log(Level.WARNING, str);
                            meta.add(new Pair<Object, Object>("error", str));
                        }
                    } catch(Exception e) {
                        String str = MessageFormat.format("Exception {0} in {1} at {2}", e, p, tick);
                        LOG.log(Level.WARNING, str);
                        LOG.log(Level.WARNING, null, e);
                        meta.add(new Pair<Object, Object>("error", str));
                    }
                    meta.add(new Pair<Object, Object>(p, list));
                    if(!complete) {
                        break;
                    }
                }
            }
            break;
            case ConsoleCmd: {
                ByteBuffer b = data;
                Level l = Level.FINE;
                String cmd = DataUtils.getText(b).trim();
                meta.add(new Pair<Object, Object>(this, cmd));
                if(b.remaining() > 0) {
                    LOG.log(l, "Underflow: {0}, {1}", new Object[] {b.remaining(), b.position()});
                }
            }
            break;
            case UserCmd: {
                // https://github.com/LestaD/SourceEngine2007/blob/master/se2007/game/shared/usercmd.cpp#L199
                BitBuffer bb = new BitBuffer(data);
                if(bb.getBoolean()) {
                    meta.add(new Pair<Object, Object>("Command number", bb.getInt()));
                } else {
                    // Assume steady increment
                }
                if(bb.getBoolean()) {
                    meta.add(new Pair<Object, Object>("Tick count", bb.getInt()));
                } else {
                    // Assume steady increment
                }
                if(bb.getBoolean()) {
                    meta.add(new Pair<Object, Object>("Viewangle pitch", bb.getFloat()));
                }
                if(bb.getBoolean()) {
                    meta.add(new Pair<Object, Object>("Viewangle yaw", bb.getFloat()));
                }
                if(bb.getBoolean()) {
                    meta.add(new Pair<Object, Object>("Viewangle roll", bb.getFloat()));
                }
                if(bb.getBoolean()) {
                    meta.add(new Pair<Object, Object>("Foward move", bb.getFloat()));
                }
                if(bb.getBoolean()) {
                    meta.add(new Pair<Object, Object>("Side move", bb.getFloat()));
                }
                if(bb.getBoolean()) {
                    meta.add(new Pair<Object, Object>("Up move", bb.getFloat()));
                }
                if(bb.getBoolean()) {
                    meta.add(new Pair<Object, Object>("Buttons", Input.get(bb.getInt())));
                }
                if(bb.getBoolean()) {
                    meta.add(new Pair<Object, Object>("Impulse", bb.getByte()));
                }
                if(bb.getBoolean()) {
                    meta.add(new Pair<Object, Object>("Weapon select", bb.getBits(HL2DEM.MAX_EDICT_BITS)));
                    if(bb.getBoolean()) {
                        meta.add(new Pair<Object, Object>("Weapon subtype", bb.getBits(HL2DEM.WEAPON_SUBTYPE_BITS)));
                    }
                }
                if(bb.getBoolean()) {
                    meta.add(new Pair<Object, Object>("Mouse Dx", bb.getShort()));
                }
                if(bb.getBoolean()) {
                    meta.add(new Pair<Object, Object>("Mouse Dy", bb.getShort()));
                }
                if(bb.remaining() > 0) {
                    meta.add(new Pair<Object, Object>("Underflow", bb.remaining()));
                }
            }
            break;
                // TODO
            case DataTables:
            case StringTables:
                break;
        }
    }

}

File: src/main/kotlin/com/timepath/hl2/io/demo/Message.kt
package com.timepath.hl2.io.demo

import com.timepath.DataUtils
import com.timepath.Logger
import com.timepath.io.BitBuffer
import com.timepath.io.OrderedOutputStream
import com.timepath.io.struct.StructField
import com.timepath.with
import java.nio.BufferUnderflowException
import java.nio.ByteBuffer
import java.text.MessageFormat
import java.util.LinkedList
import java.util.logging.Level

public class Message(
        private val outer: HL2DEM,
        public val type: MessageType?,
        StructField(index = 1) public val tick: Int) {

    public var data: ByteBuffer? = null
        set(data) {
            $data = data
            $size = data?.capacity() ?: 0
        }

    public var meta: TupleMap<Any, Any?> = TupleMap()
    public var incomplete: Boolean = false
    /** Command / sequence info. TODO: use */
    StructField(index = 2, nullable = true) public var cseq: ByteArray? = null
    /** Outgoing sequence number. TODO: use */
    StructField(index = 3, nullable = true) public var oseq: ByteArray? = null
    StructField(index = 4) public var size: Int = 0
    StructField(index = 0) private val op: Byte = 0
    private var parsed: Boolean = false

    public fun write(out: OrderedOutputStream) {
        out.writeByte(type!!.ordinal() + 1)
        out.writeInt(tick)
        cseq?.let { out.write(it) }
        oseq?.let { out.write(it) }
        when (type) {
            MessageType.Synctick, MessageType.Stop -> Unit
            else -> out.writeInt(size)
        }
        data?.let {
            it.position(0)
            val dst = ByteArray(it.limit())
            it.get(dst)
            out.write(dst)
        }
    }

    override fun toString() = "${type}, tick ${tick}, ${data?.limit() ?: 0} bytes"

    fun parse() {
        if (type == null) return
        if (data == null) return
        if (parsed) return
        when (type) {
            MessageType.Signon, MessageType.Packet -> {
                val bb = BitBuffer(data!!)
                var error: String? = null
                var thrown: Throwable? = null
                val opSize = if ((outer.header.networkProtocol >= 16)) 6 else 5
                while (bb.remainingBits() > opSize) {
                    try {
                        val op = bb.getBits(opSize).toInt()
                        val type = Packet.Type[op]
                        if (type == null) {
                            error = MessageFormat.format("Unknown message type {0} in {1}", op, this)
                            thrown = Exception("Unknown message")
                        } else {
                            val p = Packet(type, bb.positionBits())
                            try {
                                if (!type.handler.read(bb, p.list, outer)) {
                                    error = MessageFormat.format("Incomplete read of {0} in {1}", p, this)
                                }
                            } catch (e: BufferUnderflowException) {
                                error = MessageFormat.format("Out of data in {0}", this)
                            } catch (e: Exception) {
                                error = MessageFormat.format("Exception in {0} in {1}", p, this)
                                thrown = e
                            }

                            meta[p] = p.list
                        }
                    } catch (e: BufferUnderflowException) {
                        error = MessageFormat.format("Out of data in {0}", this)
                    }

                    meta["remaining bits"] = bb.remainingBits()
                    if (error != null) {
                        incomplete = true
                        meta["error"] = error
                        thrown?.let { LOG.log(Level.WARNING, { error }, it) }
                        break
                    }
                }
            }
            MessageType.ConsoleCmd -> {
                val cmd = DataUtils.getText(data!!, true)
                meta["cmd"] = cmd
            }
            MessageType.UserCmd -> {
                // https://github.com/LestaD/SourceEngine2007/blob/master/se2007/game/shared/usercmd.cpp#L199
                val bb = BitBuffer(data!!)
                if (bb.getBoolean()) {
                    meta["Command number"] = bb.getInt()
                } // else assume steady increment
                if (bb.getBoolean()) {
                    meta["Tick count"] = bb.getInt()
                } // else assume steady increment
                if (bb.getBoolean()) {
                    meta["Viewangle pitch"] = bb.getFloat()
                }
                if (bb.getBoolean()) {
                    meta["Viewangle yaw"] = bb.getFloat()
                }
                if (bb.getBoolean()) {
                    meta["Viewangle roll"] = bb.getFloat()
                }
                if (bb.getBoolean()) {
                    meta["Foward move"] = bb.getFloat()
                }
                if (bb.getBoolean()) {
                    meta["Side move"] = bb.getFloat()
                }
                if (bb.getBoolean()) {
                    meta["Up move"] = bb.getFloat()
                }
                if (bb.getBoolean()) {
                    meta["Buttons"] = Input[bb.getInt()]
                }
                if (bb.getBoolean()) {
                    meta["Impulse"] = bb.getByte()
                }
                if (bb.getBoolean()) {
                    meta["Weapon select"] = bb.getBits(HL2DEM.MAX_EDICT_BITS)
                    if (bb.getBoolean()) {
                        meta["Weapon subtype"] = bb.getBits(HL2DEM.WEAPON_SUBTYPE_BITS)
                    }
                }
                if (bb.getBoolean()) {
                    meta["Mouse Dx"] = bb.getShort()
                }
                if (bb.getBoolean()) {
                    meta["Mouse Dy"] = bb.getShort()
                }
                if (bb.remaining() > 0) {
                    meta["Underflow"] = bb.remaining()
                }
            }
            MessageType.DataTables, MessageType.StringTables -> Unit // TODO
        }
        parsed = true
    }

    companion object {

        private val LOG = Logger()

        fun parse(outer: HL2DEM, buffer: ByteBuffer): Message {
            val op = buffer.get().toInt()
            val type = MessageType[op]
            if (type == null) {
                LOG.severe { "Unknown demo message type encountered: ${op}" }
            }
            val tick = buffer.getInt()
            LOG.fine { "${type} at tick ${tick} (${buffer.position()}), ${buffer.remaining()} remaining bytes" }
            return Message(outer, type, tick).with {
                val m = this
                when {
                    m.type == MessageType.Synctick,
                    m.type == MessageType.Stop -> Unit
                    else -> {
                        when {
                            m.type == MessageType.Packet,
                            m.type == MessageType.Signon -> {
                                val dst = ByteArray(21 * 4)
                                buffer.get(dst)
                                m.cseq = dst
                            }
                            m.type == MessageType.UserCmd -> {
                                val dst = ByteArray(4)
                                buffer.get(dst)
                                m.oseq = dst
                            }
                        }
                        m.size = buffer.getInt()
                    }
                }
            }
        }
    }
}
